File: c/queue.c
#include "linked_list.c"
#include <stdbool.h>

typedef struct Queue
{
    LinkedList *data;
    Node *first;
} Queue;

Queue *get_queue(void)
{
    Queue *queue = (Queue *) malloc(sizeof(Queue));
    LinkedList *lst = get_list();
    queue->data = lst;
    queue->first = 0;
    return queue;
}

void enqueue(Queue *queue, char *value)
{
    Node *new_node = get_node(value);
    if (queue->data->length == 0)
        queue->first = new_node;
    insert(queue->data, new_node, queue->data->length);
}

Node *dequeue(Queue *queue)
{
    if (queue->data->length > 0)
    {
        Node *removed = delete_node(queue->data, 0);
        if (queue->data->length > 0)
            queue->first = queue->first->next;
        else
            queue->first = NULL;
        return removed;
    }
    return NULL;
}

bool if_empty(Queue *queue)
{
    if (queue->first != NULL)
        return false;
    return true;
}

Node *peek(Queue *queue)
{
    return queue->first;
}

void free_queue(Queue *queue)
{
    if (queue != NULL)
    {
        free_list(queue->data);
        free(queue);
    }
}

int main(void)
{
    Queue *queue = get_queue();
    enqueue(queue, "A");
    printf("%s\n", peek(queue)->value);
    print_linked_list(queue->data);
    enqueue(queue, "B"); 
    print_linked_list(queue->data);
    printf("%s\n", peek(queue)->value);
    enqueue(queue, "C");
    print_linked_list(queue->data);
    printf("%s\n", peek(queue)->value);
    dequeue(queue);
    printf("%s\n", peek(queue)->value);
    print_linked_list(queue->data);
    dequeue(queue);
    printf("%s\n", peek(queue)->value);
    print_linked_list(queue->data);
    dequeue(queue);
    print_linked_list(queue->data);
}

File: data_structures_python/queue.py
from linked_list import LinkedList, Node

class Queue:
    def __init__(self):
        self.data = LinkedList()
        self.first = None
        self.length = 0
    
    def enqueue(self, item):
        new_node = Node(item)
        if self.length == 0:
            self.first = new_node
        self.data.insert(new_node, self.length)
        self.length += 1

    def dequeue(self):
        if self.length > 0:
            removed = self.data.delete(0)
            self.length -= 1
            if self.length > 0:
                self.first = self.first.next 
            else:
                self.first = None
            return removed
        return None

    def peek(self):
        return self.first
    
    def is_empty(self):
        if self.length == 0:
            return True
        return False
    

if __name__ == "__main__":
    queue = Queue()
    queue.enqueue("first")
    queue.enqueue("second")
    queue.enqueue("third")
    print("Peek: ", queue.peek().data)
    queue.data.print_list()
    print(queue.is_empty())
    print(queue.dequeue())
    queue.data.print_list()
    print(queue.is_empty())
    print(queue.dequeue())
    queue.data.print_list()
    print(queue.is_empty())
    print(queue.dequeue())
    queue.data.print_list()
    print(queue.is_empty())





